[quant] Quantized Tensor support deepcopy

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28612 [quant] Quantized Tensor support copy**

Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

Differential Revision: [D18255247](https://our.internmc.facebook.com/intern/diff/D18255247)

jerryzh168 committed:
[quant] Quantized Tensor support copy

Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gchanan reviewed (commented):
can you remind me why quantized tensors use qint8 storages and not int8 storages?

jerryzh168 commented:
> can you remind me why quantized tensors use qint8 storages and not int8 storages?

I don't remember exactly, but one reason I can think of right now is because when we inspect the storage we want to distinguish between qint8 and int8 storage?

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gchanan reviewed (commented):
does this support copy_, i.e. `a.copy_(b)`?

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> does this support copy_, i.e. `a.copy_(b)`?

Yes it does, test added

gchanan commented on test/test_quantized_tensor.py:
```diff
@@ -332,5 +333,18 @@ def test_qscheme_pickle(self):
+        qc = deepcopy(q)
+        self.assertEqual(qc, q)
+        # copy_
+        qd = torch._empty_affine_quantized([3, 5], scale=scale, zero_point=zero_point, dtype=torch.quint8)
```
what if the scale, zero_point, dtype, type of quantization, etc are different?

jerryzh168 replied:
code is here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Copy.cpp#L118-L123
it requires same quantization scheme and dtype, but allows to copy from different scale/zero_point.

jerryzh168 replied:
these are already tested in `test_qtensor_copy` actually, I'll update the test

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Quantized Tensor support copy"


Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@gchanan this PR only adds support for deepcopy, normal quantized tensor copy is already supported in another PR: https://github.com/pytorch/pytorch/pull/21852

gchanan reviewed (approved):
this, in-and-of-itself, looks fine for deepcopy.

Why was the existing copy behavior (of clobbering the quantizer of the destination) chosen?  That seems surprising, and I don't know why you even want to copy in that case.

jerryzh168 commented:
> this, in-and-of-itself, looks fine for deepcopy.
> 
> Why was the existing copy behavior (of clobbering the quantizer of the destination) chosen? That seems surprising, and I don't know why you even want to copy in that case.

It's because of: https://github.com/pytorch/pytorch/blob/e5d6b753192cde81148d812ad52426e51e89e903/torch/nn/modules/module.py#L771
